bazel: fix strict action env

guibou opened the pull request:
##### Motivation for this change

This fix three bugs:

- ~~it restores the usage of `/etc/bazel.bazelrc` if users want to override bazel setting on a system-wide fashion.~~
- it removes `/usr/local/bin` from the path searched by bazel in "strict action env" mode. This increases reproducibility in this context.
- it adds `diff` and `zip` in the bazel test runner context. They are needed on some cases of test and where failing if bazel was used with `strict action env`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @flokli @Profpatsch

guibou committed:
bazel: improve strict action env context

- Bazel strict action env set a default PATH to
`/bin/:/usr/bin:/usr/local/bin`. This was previously changed to disable
this behavior to improve hermeticity. However the previous change was
only removing `/bin:/usr/bin`, keeping `/usr/local/bin`, this commit
also remove this entry.

jonringer reviewed (commented):
seems to break a lot, then again, I think most of this is broken to begin with:

```
[20 built (10 failed), 474 copied (3168.7 MiB), 1114.8 MiB DL]
error: build of '/nix/store/clkxx1l9zpkk14hrpa36wv79jw2yfnbd-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/73309
1 package are marked as broken and were skipped:
bazel-deps

27 package failed to build:
libtensorflow python27Packages.baselines python27Packages.dm-sonnet python27Packages.edward python27Packages.graph_nets python27Packages.tensorflow python27Packages.tensorflow-probability python27Packages.tensorflowWithCuda python27Packages.tflearn python37Packages.baselines python37Packages.dm-sonnet python37Packages.edward python37Packages.graph_nets python37Packages.optuna python37Packages.rl-coach python37Packages.tensorflow python37Packages.tensorflow-probability python37Packages.tensorflowWithCuda python37Packages.tflearn python38Packages.baselines python38Packages.dm-sonnet python38Packages.edward python38Packages.graph_nets python38Packages.tensorflow python38Packages.tensorflow-probability python38Packages.tensorflowWithCuda python38Packages.tflearn

2 package were build:
bazel bazel-watcher
```

guibou commented:
@jonringer Thank you, I'm investigating it.

What exactly do you mean with:

> then again, I think most of this is broken to begin with

I have decided to spend a few time this week to improve bazel quality on nixpkgs, so if you have some ideas about what can be improved, please tell me.

jonringer commented:
tensorflow is broken, and a lot of packages rely on it

jonringer commented:
see https://github.com/NixOS/nixpkgs/pull/71282

flokli commented on pkgs/development/tools/build-managers/bazel/default.nix:
```diff
@@ -396,7 +397,7 @@ stdenv.mkDerivation rec {
       substituteInPlace \
         src/main/java/com/google/devtools/build/lib/bazel/rules/BazelRuleClassProvider.java \
-        --replace /bin:/usr/bin ${defaultShellPath}
+        --replace /bin:/usr/bin:/usr/local/bin ${defaultShellPath}
```
This patches the following line:
https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/bazel/rules/BazelRuleClassProvider.java#L472

I assume there once was only `/bin:/usr/bin` in there.

To avoid silent regressions in the future, we should use a patch replacing this with a placeholder, and substitute the placeholder before applying the patch, as done in `https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/udisks/2-default.nix#L22`.

guibou replied:
Done.

guibou commented:
@jonringer I do apologize for the build issues you have with bazel 1.0, I did the upgrade without ensuring that dependencies are working. I won't do that issue again. Eventually we'll solve `tensorflow` build, but I think that's unrelated to the current pull request.

@flokli I applied the changes you suggested.

flokli commented on pkgs/development/tools/build-managers/bazel/default.nix:
```diff
@@ -145,6 +146,10 @@ stdenv.mkDerivation rec {
+
+    # replace hard coded non hermetic paths by NIX_DEFAULT_BAZEL_SHELL_PATH
+    # we are doing this with a patch to ensure we systematically catch changes in this line
+    ./strict_action_env.patch
```
please call the placeholder `@strictActionEnvPath@` (or similar), and do a `substituteAll` on the patch file before applying it, instead of the substitution in `postPatch`. You might want to move the comment from down there to here:

```
      # --experimental_strict_action_env (which will soon become the
      # default, see bazelbuild/bazel#2574) hardcodes the default
      # action environment to a value that on NixOS at least is bogus.
      # So we hardcode it to something useful.
```

Profpatsch commented:
@GrahamcOfBorg build bazel.tests

flokli commented on pkgs/development/tools/build-managers/bazel/default.nix:
```diff
@@ -479,6 +484,10 @@ stdenv.mkDerivation rec {
 
+    # we set a custom bazelrc in $out/etc/bazelrc
+    # the following "try-import" restores the read of the system-wide bazel.bazelrc
+    echo "try-import /etc/bazel.bazelrc" >> $out/etc/bazelrc
```
Maybe a stupid question, but can we stop using `$out/etc/bazelrc` entirely, and instead add `--override_repository=${remote_java_tools.name}=${remote_java_tools}"` and `--distdir=${distDir}` to the bazel wrapper already present at https://github.com/NixOS/nixpkgs/pull/73309/files#diff-4746b81564a01e241a84d455575437e7R478?

flokli replied:
Well, plus that part:
```
      build --distdir=${distDir}
      fetch --distdir=${distDir}
      build --copt="$(echo $NIX_CFLAGS_COMPILE | sed -e 's/ /" --copt="/g')"
      build --host_copt="$(echo $NIX_CFLAGS_COMPILE | sed -e 's/ /" --host_copt="/g')"
      build --linkopt="-Wl,$(echo $NIX_LDFLAGS | sed -e 's/ /" --linkopt="-Wl,/g')"
      build --host_linkopt="-Wl,$(echo $NIX_LDFLAGS | sed -e 's/ /" --host_linkopt="-Wl,/g')"
      build --host_javabase='@local_jdk//:jdk'   
      build --host_java_toolchain='${javaToolchain}'
```

and dropping the
```
      # bazel reads its system bazelrc in /etc
      # override this path to a builtin one
      substituteInPlace \
        src/main/cpp/option_processor.cc \
        --replace BAZEL_SYSTEM_BAZELRC_PATH "\"$out/etc/bazelrc\""
```

flokli replied:
It'd probably also make sense to separate the `strict_action_env` fixes from the bazelrc improvements - so they should at least be in 2 separate commits.

guibou replied:
> Maybe a stupid question, but can we stop using `$out/etc/bazelrc` entirely, and instead add `--override_repository=${remote_java_tools.name}=${remote_java_tools}"` and `--distdir=${distDir}` to the bazel wrapper already present at https://github.com/NixOS/nixpkgs/pull/73309/files#diff-4746b81564a01e241a84d455575437e7R478?

Not possible. You can see that the options are in the `build` configuration section. However the wrapper does not know that we are using the `build` or another command. Moving this to the wrapper will need more logic, such as a detection of the command used by the bazel user.

The second part exists just at build time, that's the bazelrc of the buildtime bazel.

I do agree with the split. Done.

flokli replied:
Sorry for not being clear. I meant to split the whole bazelrc stuff in a followup PR.

guibou committed:
bazel: add `file` and `zip` in the test runner env

`file` and `zip` are needed for some bazel test and by default the test
runner take these binaries from the current `PATH` which may not contain
them

Profpatsch commented:
@GrahamcOfBorg build bazel.tests

guibou commented:
@flokli : as discussed in a private conversation, I dropped the configuration file commit, so this PR only focus on bazel environment with strict action env.

flokli commented:
@GrahamcOfBorg build bazel.tests

flokli commented:
This seems to fail on darwin:

```
src/main/tools/process-wrapper-legacy.cc:58: "execvp(external/local_config_cc/cc_wrapper.sh, ...)": No such file or directory
INFO: Elapsed time: 10.199s, Critical Path: 0.39s
INFO: 1 process: 1 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

@uri-canva @jonringer does that ring a bell?

uri-canva commented:
Yep, definitely seen it before, I'm pretty sure it's already broken on master, note it's in `bazel.tests`, not in `bazel`, so it doesn't prevent the nix binary cache from having a darwin bazel build, that's why we haven't focused on figuring that out.

flokli commented:
In that case, I don't see any strong reasons not to merge this in.
